src/viam/examples/modules/complex/main.cpp

```diff
@@ -63,6 +63,10 @@ int main(int argc, char** argv) {
 
     std::vector<std::shared_ptr<ModelRegistration>> mrs = {mybase_mr, mygizmo_mr, mysummation_mr};
     auto my_mod = std::make_shared<ModuleService>(argc, argv, mrs);
+    // register_cleanup_function can be used to run custom logic at the time of
+    // shutdown. It must be called before `serve`.
+    my_mod->register_cleanup_function(
+        []() { BOOST_LOG_TRIVIAL(info) << "C++ complex module example cleaning up"; });
```

Comment:
So, what can this cleanup function do that couldn't be done by re-writing the code like this:

```
{
    auto my_mod = std::make_shared<ModuleService>(argc, argv, mrs);
    my_mod->serve();
    // cleanup logic while my_mod still exists
}
```
Is there something that a registered cleanup function can do by virtue of executing "within" `my_mod`'s dtor that also can't just be written as ordinary code after `serve` returns and before `my_mod` is destroyed?

I'll note also that the cleanup function can really only have access to global state, other state in main, and the `ModuleService` itself. What functionality is enabled by having it that we can't write today by other means?

Reply:
> Is there something that a registered cleanup function can do by virtue of executing "within" my_mod's dtor that also can't just be written as ordinary code after serve returns and before my_mod is destroyed?

I don't think it can do anything "special" for the exact reasons that you note (no access to private state of the `ModuleService` or anything that code at the end of a `main` function wouldn't have). Perhaps the only thing I could imagine is passing the signal received (SIGINT/SIGTERM/something else) to the cleanup function so the module author could do something special based on what signal was received? If that's not desirable/useful, though, I think I should probably close this PR.

Reply:
Sure, let's close for now (we can always re-open it), and maybe move discussion over to [RSDK-5590](https://viam.atlassian.net/browse/RSDK-5509)? I don't actually want to drop the issue because I believe there is work we want to do in this space, and I think the ticket summary points in the direction of these sorts of activities:
- Registering a `terminate` handler that provides more diagnostics (exception introspection, backtracing) on unexpected shutdown via https://en.cppreference.com/w/cpp/error/set_terminate.
- Registering synchronous signal handlers for the common fatal signals (SIGSEGV, SIGILL, SIGBUS).
- Stack unwinding / backtracing support
- Allowing module authors to register their own custom async signal handlers with the signal manager (e.g. hooking SIGUSR1 to dump diagnostic data)

Reply:
Cool; moved discussion back to the ticket.